feat: remove ViewEncapsulation.Native

crisbeto opened the pull request:
See the individual commits for more context, but the high-level overview is:
1. Adds an automated migration that will replace any usages of `ViewEncapsulation.Native` with `ViewEncapsulation.ShadowDom`.
2. Removes the `Native` value from `ViewEncapsulation` and replaces any usages in tests and docs with `ShadowDom`.

crisbeto committed:
feat(core): add automated migration to replace ViewEncapsulation.Native

Adds an automated migration that replaces any usages of the deprecated
`ViewEncapsulation.Native` with `ViewEncapsulation.ShadowDom`.

IgorMinar commented on packages/core/schematics/migrations/native-view-encapsulation/index.ts:
```diff
@@ -0,0 +1,52 @@
+  const {program} = createMigrationProgram(tree, tsconfigPath, basePath);
+  const typeChecker = program.getTypeChecker();
+  const sourceFiles = program.getSourceFiles().filter(
+      f => !f.isDeclarationFile && !program.isSourceFileFromExternalLibrary(f));
```
Is it necessary to exclude declaration files?

I think it's rare that this symbol would be present in d.ts files, but if we was there, then we'd still want to replace it.

crisbeto replied:
It's something we do on all migrations. I believe we started doing it so that we don't have to go through the AST for things like third-party libraries. @devversion should know more about it.

devversion replied:
Yeah, we don't want to run on non source-files as migrations only run once, and changes to the `node_modules` wouldn't be persistent.  For excluding declaration files, that is a good question here. I think `ViewEncapsulation.Native` could be referenced in there, as a qualified name. 

We probably would want to skip such cases as they are rather unlikely and would require more code for this schematic? i.e. we would need to handle a `ts.QualifiedName` and not only the property access as currently done.

IgorMinar commented on packages/core/schematics/migrations/native-view-encapsulation/README.md:
```diff
@@ -0,0 +1,31 @@
+## `ViewEncapsulation.Native` migration
+
+Automatically migrates usages of `ViewEncapsulation.Native` to `ViewEncapsulation.ShadowDom`.
```
maybe mention that "For most practical purposes the `Native` mode is compatible with the `ShadowDom` mode."

IgorMinar reviewed (approved):
LGTM with some suggestions. thanks!

AndrewKushnir reviewed (commented):
The changes look good, just left 1 minor comment 👍

devversion reviewed (approved):
LGTM for the migration code.

crisbeto commented:
I've addressed the latest set of feedback and added a temporary fallback for code that is still using the `Native` encapsulation.

AndrewKushnir commented on packages/core/src/view/util.ts:
```diff
@@ -230,7 +230,10 @@ export function getParentRenderElement(view: ViewData, renderHost: any, def: Nod
+         (renderParent.element!.componentRendererType!.encapsulation ===
+              ViewEncapsulation.ShadowDom ||
+          // TODO: remove the Native fallback logic in v12.
+          renderParent.element!.componentRendererType!.encapsulation === 1))) {
```
I'd propose to introduce an internal constant to refer to the deprecated view encapsulation value, so that it's easier to find and remove all usages in v12 (and also keep the `TODO` comment):
```suggestion
          renderParent.element!.componentRendererType!.encapsulation === ɵDeprecatedNativeViewEncapsulation))) {
```

What do you think?

crisbeto replied:
My reasoning behind not having a constant is that we only use it in a couple of places in two different packages (`core` and `platform-browser`). Adding an exposing the constant would be a lot of boilerplate for not much benefit.

AndrewKushnir replied:
Got it, thanks for the comment. I guess in this case we can introduce separate constants (with the same name) in two packages (so we don't have to link them together). The main benefit is to have an easy way to find all references later (and it'd be convenient to use a symbol for it), so we can clean them up. Another approach might be to add an inline comment like this:

```
          // TODO: remove the Native fallback logic in v12.
          renderParent.element!.componentRendererType!.encapsulation === 1 /* deprecated ViewEncapsulation.Native */))) {
```

crisbeto replied:
I left a comment next to them, but it might be best to create a Jira task and link it in the comment. I'll do that when addressing the rest of the feedback.

AndrewKushnir replied:
Ok, thanks 👍

AndrewKushnir commented on packages/platform-browser/src/dom/dom_renderer.ts:
```diff
@@ -100,8 +102,15 @@ export class DomRendererFactory2 implements RendererFactory2 {
+        if (!hasLoggedNativeEncapsulationWarning && type.encapsulation === 1) {
+          hasLoggedNativeEncapsulationWarning = true;
+          console.warn(
+              'ViewEncapsulation.Native is no longer supported. Falling back to ShadowDom. The fallback will be removed after v12.');
```
```suggestion
              'ViewEncapsulation.Native is no longer supported. Falling back to ViewEncapsulation.ShadowDom. The fallback will be removed in v12.');
```

May be we can also use `ngDevMode` here to tree-shake away this error message for prod bundles?

AndrewKushnir replied:
I think CLI uses `ngDevMode` for both VE and Ivy (for AOT mode), we'd just need to make sure we are not accessing a global that was not defined (for JIT cases), i.e. `typeof ngDevMode !== 'undefined' && ...`.

AndrewKushnir reviewed (commented):
Thanks for making updates @crisbeto, I just left a couple comments.

crisbeto commented:
I've addressed the latest set of feedback @AndrewKushnir.

AndrewKushnir reviewed (approved):
LGTM, thanks for additional updates @crisbeto 👍

AndrewKushnir commented:
[Presubmit](http://test/OCL:334435354:BASE:334435374:1601407212398:6d4ce01c).

AndrewKushnir reviewed (approved):
Reviewed-for: public-api

atscott reviewed (approved):
reviewed-for: public-api

petebacondarwin commented on goldens/size-tracking/integration-payloads.json:
```diff
@@ -30,7 +30,7 @@
       "uncompressed": {
         "runtime-es2015": 1485,
-        "main-es2015": 134891,
+        "main-es2015": 135003,
```
How did removing something make the size go up?

petebacondarwin commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -517,8 +517,8 @@ function getRenderParent(tView: TView, tNode: TNode, currentView: LView): REleme
-      if (encapsulation !== ViewEncapsulation.ShadowDom &&
-          encapsulation !== ViewEncapsulation.Native) {
+      // TODO(FW-2290): remove the `encapsulation !== 1` fallback logic in v12.
+      if (encapsulation !== ViewEncapsulation.ShadowDom && encapsulation !== 1) {
```
Why note use?

```suggestion
      if (encapsulation === ViewEncapsulation.None || encapsulation === ViewEncapsulation.Emulated) {
```

Then you don't need to worry about removing the `1` later.

petebacondarwin replied:
I think this would also be preferred since we can't be sure that adding a new `ViewEncapsulation` type in the future would also allow this early escape.

petebacondarwin commented on packages/platform-browser/testing/src/matchers.ts:
```diff
@@ -305,8 +305,14 @@ function elementText(n: any): string {
 
-  if (getDOM().isElementNode(n) && (n as Element).tagName == 'CONTENT') {
-    return elementText(Array.prototype.slice.apply((<any>n).getDistributedNodes()));
+  if (getDOM().isElementNode(n)) {
```
As an aside, it is a shame that `isElementNode()` is not a type guard.

petebacondarwin reviewed (commented):
Some drive-by comments that you can ignore.

petebacondarwin reviewed (approved):
Reviewed-for: public-api

petebacondarwin reviewed (approved):
Reviewed-for: size-tracking

AndrewKushnir reviewed (approved):
Reviewed-for: size-tracking

crisbeto committed:
feat(core): remove ViewEncapsulation.Native

Removes `ViewEncapsulation.Native` which has been deprecated for several major versions.

BREAKING CHANGES:
* `ViewEncapsulation.Native` has been removed. Use `ViewEncapsulation.ShadowDom` instead. Existing
usages will be updated automatically by `ng update`.
